pkg/vulnsrc/vulnsrc.go

```diff
@@ -61,53 +63,88 @@ func init() {
 	}
 }
 
-func Update(targets []string, cacheDir string, light bool, updateInterval time.Duration) error {
+type Operation interface {
+	SetMetadata(db.Metadata) error
+}
+
+type Updater struct {
+	dbc            Operation
+	updateMap      map[string]VulnSrc
+	cacheDir       string
+	dbType         db.Type
+	updateInterval time.Duration
+	clock          clock.Clock
+	optimizer      Optimizer
+}
+
+func NewUpdater(cacheDir string, light bool, interval time.Duration) Updater {
+	var optimizer Optimizer
+	dbConfig := db.Config{}
+	dbType := db.TypeFull
+	optimizer = fullOptimizer{dbc: dbConfig}
+
+	if light {
+		dbType = db.TypeLight
+		optimizer = lightOptimizer{dbc: dbConfig}
+	}
+
+	return Updater{
+		dbc:            db.Config{},
+		updateMap:      updateMap,
+		cacheDir:       cacheDir,
+		dbType:         dbType,
+		updateInterval: interval,
+		clock:          clock.RealClock{},
+		optimizer:      optimizer,
+	}
+}
+
+func (u Updater) Update(targets []string) error {
 	log.Println("Updating vulnerability database...")
 
 	for _, distribution := range targets {
-		vulnSrc, ok := updateMap[distribution]
+		vulnSrc, ok := u.updateMap[distribution]
 		if !ok {
 			return xerrors.Errorf("%s does not supported yet", distribution)
 		}
 		log.Printf("Updating %s data...\n", distribution)
 
-		if err := vulnSrc.Update(cacheDir); err != nil {
+		if err := vulnSrc.Update(u.cacheDir); err != nil {
 			return xerrors.Errorf("error in %s update: %w", distribution, err)
 		}
 	}
 
-	dbc := db.Config{}
-	dbType := db.TypeFull
-	if light {
-		dbType = db.TypeLight
-	}
-
-	err := dbc.SetMetadata(db.Metadata{
+	err := u.dbc.SetMetadata(db.Metadata{
 		Version:    db.SchemaVersion,
-		Type:       dbType,
-		NextUpdate: time.Now().UTC().Add(updateInterval),
-		UpdatedAt:  time.Now().UTC(),
+		Type:       u.dbType,
+		NextUpdate: u.clock.Now().UTC().Add(u.updateInterval),
+		UpdatedAt:  u.clock.Now().UTC(),
 	})
 	if err != nil {
 		return xerrors.Errorf("failed to save metadata: %w", err)
 	}
 
-	if light {
-		return optimizeLightDB(dbc)
-	}
-	return optimizeFullDB(dbc)
+	return u.optimizer.Optimize()
 }
 
-func optimizeFullDB(dbc db.Config) error {
-	err := dbc.ForEachSeverity(func(tx *bolt.Tx, cveID string, _ types.Severity) error {
+type Optimizer interface {
+	Optimize() error
+}
+
+type fullOptimizer struct {
+	dbc db.Operations
+}
+
+func (o fullOptimizer) Optimize() error {
```

Comment:
Can we include a test for both `Optimize()` versions, light and full? Currently they remain untested.

Reply:
My bad. This has been covered https://github.com/aquasecurity/trivy-db/pull/20